### DISCUSSION TITLE
Printout of Jacobian/Residual

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#22961] [Q&A Modules: General] [answered]
I am trying to debug an app.  The contributions to the residuals from my kernels appear to be correct.  I am using AD to form the jacobian.  I did run  analyzejacobian and got "No errors detected.  :-)".  [of course, thanks to the AD system]

But something must not be quite right as I don't get the same results as for a hand calculation. I have seen a Moose developer show a printout of the jacobian for a small test problem (see below).  How can I do this? 

      SVD: condition number 2.441843322518e+11, 0 of 9 singular values are (nearly) zero
      SVD: singular values:
             1.501529298631e+06   1.038973912264e+06   3.701408259763e+05   4.668696900287e+01   1.192502051978e+01  
             2.869699441978e+00   8.516552976210e-06   7.377570511006e-06   6.149163153854e-06  
      0 Linear |R| = 2.824152e-04
      1 Linear |R| = 4.153350e-13
Mat Object: () 1 MPI process
  type: seqaij
row 0: (0, 583.966)  (1, 0.000186027)  (2, 0.000510781)  (3, -581.157)  (4, -0.000186357)  (5, 0.499837) 
row 1: (0, 4221.37)  (1, 0.00139813)  (2, 0.0180354)  (3, -4224.11)  (4, -0.00135452)  (5, 3.63304) 
row 2: (0, -833398.)  (1, -0.267009)  (2, 1437.37)  (3, 832664.)  (4, 0.267006)  (5, -716.151) 
row 3: (0, -581.162)  (1, -0.000186357)  (2, -0.500163)  (3, 1165.65)  (4, 0.000372714)  (5, 0.00032579)  (6, -581.154)  (7, -0.000186357)  (8, 0.499837) 
row 4: (0, -4233.51)  (1, -0.00136985)  (2, -3.64972)  (3, 8457.57)  (4, 0.00275771)  (5, 0.0166714)  (6, -4224.07)  (7, -0.00135452)  (8, 3.63302) 
row 5: (0, -833558.)  (1, -0.26775)  (2, -717.614)  (3, 883.188)  (4, 0.000743158)  (5, 1437.09)  (6, 832654.)  (7, 0.267005)  (8, -716.147) 
row 6: (3, -581.157)  (4, -0.000186357)  (5, -0.500163)  (6, 1165.14)  (7, 0.000372715)  (8, 0.499976) 
row 7: (3, -4233.46)  (4, -0.00136985)  (5, -3.64971)  (6, 8453.85)  (7, 0.0027577)  (8, 3.64832) 
row 8: (3, -833547.)  (4, -0.267749)  (5, -717.609)  (6, 1600.23)  (7, 0.000743599)  (8, 721.21) 

Thanks,
  Joe Kelly

**Comment by GiudGiud** [comment, marked as answer]
Hi Joe

The flags at play here:
The svd monitor for the condition number
```
petsc_options = '-pc_svd_monitor'
petsc_options_iname = '-pc_type'
petsc_options_value = 'svd'
```
and this other petsc_option(s) for the matrix printout (you must use Newton method for that one, use pmat for PJFNK)
`-ksp_view_mat`

Guillaume

**Reply by joe61vette** [reply]
THANKS!!  That worked great.  But could you please explain a little.  I understand the jacobian matrix and it matches what I expected.  But what does the 'condition number' mean?  Mine is 10^21.

Also, what are the 'singular values'?   I get:

SVD: condition number 7.113204853455e+20, 3 of 7 singular values are (nearly) zero
SVD: singular values: 1.623154264692e-21
 SVD: singular values: 8.796004189038e-01

And finally, with 

  solve_type = 'NEWTON'
  petsc_options = '-pc_svd_monitor -ksp_view_mat'
  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_shift_amount'
  petsc_options_value = 'svd              NONZERO               1e-10'

where svd replaced lu in petsc_options_value, my test problem CONVERGED!  Whereas I was getting a linear iteration failure.  Not only did it converge, it gave the correct answer.  So, what should my pets options be?

Thanks again,
  Joe

**Reply by GiudGiud** [reply]
~~The condition number is the ratio of the first two largest eigenvalues (or smallest, I m likely getting confused with neutronics keff right here, no time to google I have a meeting) of the matrix.~~ see below

10^21 is very severely ill conditioned. We can get that down a little bit with scaling often, but at these heights it s usually a modeling issue like a missing boundary condition creating a null space. 
The number of singular values is often tied to the number of dofs that are not well-posed.

svd is more of a debugging tool, it can barely be used (extremely slow) above 1k dofs.

Moving forward if the matrix is only 7 by 7, the matrix print out should let you examine it in matlab or python. There is something wrong with it, making the inversion difficult. @lindsayad do you know of some resources for that?

**Reply by joe61vette** [reply]
Looking at it now.  There is something definitely wrong with the matrix.  Lot's for me to understand here as I am trying a hybrid rDG/CG scheme.  So velocity is essentially FVM while pressure is CG.  So, a 3 element model has 3 velocities and 4 pressures.

Thanks so much Guillaume.  I'll mark this as answered.

**Reply by lindsayad** [reply]
The condition number is the ratio of largest and smallest singular values

**Reply by lindsayad** [reply]
Since we always use the same matrix for the Jacobian and preconditioner, the `-ksp_view_pmat` option can be used for both NEWTON and PJFNK

**Reply by lindsayad** [reply]
I would be worried about [LBB stability](https://en.wikipedia.org/wiki/Ladyzhenskaya%E2%80%93Babu%C5%A1ka%E2%80%93Brezzi_condition) of a scheme in which you have more pressure dofs than velocity dofs. For incompressible flow the pressure is a Lagrange multiplier enforcing the divergence free constraint ... if you have more constraint dofs than primal dofs then I could easily see producing linearly dependent rows in your matrix

**Reply by GiudGiud** [reply]
oh yeah my bad I was thinking of the dominance ratio for some reason.

**Reply by joe61vette** [reply]
The test problem is a simple one.  Three elements & 1 D single-phase.  No flow, just a stagnant liquid column to check pressure and gravity kernels.  Two variables, Vx and p.  Vx is a constant monomial at the cell-center.  P is first order gauss. So there are 4 nodes and hence 4 pressures. This is the hybrid scheme described in the paper Alex referenced.  But one of the pressures is at the outlet and hence is covered by a Dirichlet bc.  So, I have 2 eqs x 3 elements and 6 unknowns.

There is a problem with my jacobian.  I must have a mistake somewhere in my rDG flux kernels.  Although the contribution of the mass & momentum advection terms to the element/neighbor residuals are correct and have their correct AD derivatives, somehow those terms don’t show up in the jacobian.  Indeed I have a row that is all zeros.  No wonder lu couldn’t solve it.  I am amazed that svd (whatever it is) can solve it.

I have an incompressible fluid, so the mass eq residuals are only fns of the velocity.  For momentum, due to zero velocity, the advection terms are zero and so the momentum residual is only a fn of the pressure gradient.  In other words, no diagonal terms whatsoever. Pretty severe test.  Lots of meetings today but hopefully some time to figure out why the derivatives of the mass eq residuals (advection so deriv is just the density) don’t make it into jacobian.


Thanks for the help.  I like the idea of this hybrid scheme though making it 2nd order for a 3D problem will take a lot of work.  For now, simple 1st order upwind.

One more question, is it possible to printout the Residual vector?  Thanks.

**Reply by GiudGiud** [reply]
I think it's possible to print the residual vector using vector tagging. I ll dig out an example

**Reply by joe61vette** [reply]
Thanks Guillaume.  Small correction in the above. My pressure variable is First/Lagrange.  It is the quadrature that is First/Gauss.

**Reply by GiudGiud** [reply]
You ll want to use the tagging interface
https://mooseframework.inl.gov/source/interfaces/TaggingInterface.html

The syntax will look like this:
```
[AuxVariables]
  [an_aux]
  []
[]

[AuxKernels]
  [tag_output]
    type = TagVectorAux
    v = u
    variable = an_aux
    vector_tag = 'tag_diff'
  []
[]

[Kernels]
  [diff]
    type = Diffusion
    variable = u
    extra_vector_tags = 'tag_diff'
  []
[]

[Problem]
  extra_tag_vectors  = 'tag_diff'
[]
```

I think by replacing vector by matrix you can get the  jacobian contribution

**Reply by GiudGiud** [reply]
If you specify the same tag to multiple kernels, you ll get the sum of the contributions in that tag btw

**Reply by joe61vette** [reply]
It looks like I will be needing to go into ADDGKernal to try to understand why the derivatives of my residuals are not being correctly inserted into the jacobian.  I am a bit afraid that it might be due to the usage of a hybrid scheme.  Namely, for the mass advection residual, the variable is pressure, and it is evaluated twice (for two values of _i).  Whereas for momentum advection, the variable is velocity, and it is evaluated only once (_i is equal to 0 only).

**Reply by lindsayad** [reply]
You can also do `-snes_monitor_residual` if you're interested in viewing the whole vector

**Reply by lindsayad** [reply]
@joe61vette can you share your input file?

**Reply by joe61vette** [reply]
@lindsayad I'll email you the input file.  For the kernels etc. look at the ASP repo, the rDG branch.  Just pushed it, so there are a lot of debug print statements in it.

**Reply by joe61vette** [reply]
BTW, any advice on how to implement a hybrid scheme is very welcome!

**Reply by joe61vette** [reply]
@lindsayad You'll note that I have not yet implement rDG flux Bus for inlet & outlet.  Just trying to figure out how things get calculated before trying to solve flow problems.  Gravity/pressure was the simplest I could come up with.  Somehow, svd actually gets the correct gravity head while keeping velocities zero.

**Reply by lindsayad** [reply]
So there is a fundamental problem when trying to run a DGKernel on a continuous variable, which is what you are trying to do with the `MassAdvection` DGKernel. From the elem side you will get derivatives like 
```
499.10500000000002, 499.10500000000002
```
in the computed residual. And from the neighbor side you will get
```
-499.1049999999999, -499.1049999999999
```
because you are just applying a minus sign to the flux. And you are summing these together for the exact same dof index for pressure because you are running right on the pressure node. This is why you are getting all singular rows for the pressure dofs (other than the Dirichlet row)

**Reply by lindsayad** [reply]
I would assume in a hybrid scheme in which pressure is a Lagrange variable that all kernels for the pressure variable should be regular kernels, e.g. no DGKernels with `variable = p`

**Reply by joe61vette** [reply]
Thanks Alex.  That makes perfect sense. I’ll have to think about how to treat the mass advection then.  Using the simple upwind treatment with a DGKernel was really nice.

Also, once I go to two-phase, one mass kernel will have p as the variable while the other has phasic volume fraction.
